Invalid suggestion to add dyn to something that is not a trait

### Code

```Rust
fn derp<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
    x()
}
```

### Current output

```Shell
Checking foo v0.1.0 (/home/pacak/tmp/foo)
error[E0404]: expected trait, found builtin type `usize`
 --> src/lib.rs:1:39
  |
1 | fn derp<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
  |                                       ^^^^^ not a trait

error[E0782]: expected a type, found a trait
 --> src/lib.rs:1:39
  |
1 | fn derp<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
  |                                       ^^^^^^^^^^
  |
help: you can add the `dyn` keyword if you want a trait object
  |
1 | fn derp<'a>(x: Box<dyn Fn() -> Option<dyn usize + 'a>>) -> usize {
  |                                       +++

Some errors have detailed explanations: E0404, E0782.
For more information about an error, try `rustc --explain E0404`.
error: could not compile `foo` (lib) due to 2 previous errors
```

### Desired output

Not suggesting `dyn` would be a good start, suggesting to move `>` to the left would be even better

### Rationale and extra context

Let's start with some valid code (lifetime is redundant here, but makes sense in my full code).

```rust
fn derp<'a>(x: Box<dyn Fn() -> usize + 'a>) -> usize {
    x(); todo!();
}
```

And change it to the value produced is `Optional`:

This works:
```rust
fn derp<'a>(x: Box<dyn Fn() -> Optional<usize> + 'a>) -> usize {
    x(); todo!();
}
```

But imagine making a typo...
```rust
fn derp<'a>(x: Box<dyn Fn() -> Optional<usize + 'a>>) -> usize {
    x(); todo!();
}
```

Error message is confusing and suggestion is wrong.

### Other cases


Nightly suggests to add a parameter for a closure that takes no parameters... In addition to suggesting to adding `dyn`

```Console
error[E0057]: this function takes 1 argument but 0 arguments were supplied
 --> src/lib.rs:2:5
  |
2 |     x()
  |     ^-- argument #1 is missing
  |
note: implementation defined here
 --> /rustc/f8c27dfe1a2e7fb538fd91dad53de06992c7c967/library/alloc/src/boxed.rs:1978:1
help: provide the argument
  |
2 |     x(/* value */)
  |       +++++++++++
```

### Rust Version

```Shell
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: x86_64-unknown-linux-gnu
release: 1.85.0
LLVM version: 19.1.7
```

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"xizheyin"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
@@ -86,6 +86,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 "expected a type, found a trait"
             );
             if self_ty.span.can_be_used_for_suggestions()
+                && poly_trait_ref.trait_ref.trait_def_id().is_some()
                 && !self.maybe_suggest_impl_trait(self_ty, &mut diag)
                 && !self.maybe_suggest_dyn_trait(self_ty, sugg, &mut diag)
             {

--- a/tests/ui/traits/object/suggestion-trait-object-issue-139174.rs
+++ b/tests/ui/traits/object/suggestion-trait-object-issue-139174.rs
@@ -1,0 +1,24 @@
+//@compile-flags: --edition 2021
+
+fn f<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
+    //~^ ERROR expected trait, found builtin type `usize`
+    //~| ERROR expected a type, found a trait [E0782]
+    0
+}
+
+fn create_adder<'a>(x: i32) -> usize + 'a {
+    //~^ ERROR expected trait, found builtin type `usize`
+    //~| ERROR expected a type, found a trait [E0782]
+    move |y| x + y
+}
+
+struct Struct<'a>{
+    x: usize + 'a,
+    //~^ ERROR expected trait, found builtin type `usize`
+    //~| ERROR expected a type, found a trait [E0782]
+}
+
+
+fn main() {
+
+}

--- a/tests/ui/traits/object/suggestion-trait-object-issue-139174.stderr
+++ b/tests/ui/traits/object/suggestion-trait-object-issue-139174.stderr
@@ -21,33 +21,18 @@ error[E0782]: expected a type, found a trait
    |
 LL | fn f<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
    |                                    ^^^^^^^^^^
-   |
-help: you can add the `dyn` keyword if you want a trait object
-   |
-LL | fn f<'a>(x: Box<dyn Fn() -> Option<dyn usize + 'a>>) -> usize {
-   |                                    +++
 
 error[E0782]: expected a type, found a trait
   --> $DIR/suggestion-trait-object-issue-139174.rs:9:32
    |
 LL | fn create_adder<'a>(x: i32) -> usize + 'a {
    |                                ^^^^^^^^^^
-   |
-help: `usize + 'a` is dyn-incompatible, use `impl usize + 'a` to return an opaque type, as long as you return a single underlying type
-   |
-LL | fn create_adder<'a>(x: i32) -> impl usize + 'a {
-   |                                ++++
 
 error[E0782]: expected a type, found a trait
   --> $DIR/suggestion-trait-object-issue-139174.rs:16:8
    |
 LL |     x: usize + 'a,
    |        ^^^^^^^^^^
-   |
-help: you can add the `dyn` keyword if you want a trait object
-   |
-LL |     x: dyn usize + 'a,
-   |        +++
 
 error: aborting due to 6 previous errors
 

--- a/tests/ui/traits/object/suggestion-trait-object-issue-139174.stderr
+++ b/tests/ui/traits/object/suggestion-trait-object-issue-139174.stderr
@@ -1,0 +1,55 @@
+error[E0404]: expected trait, found builtin type `usize`
+  --> $DIR/suggestion-trait-object-issue-139174.rs:3:36
+   |
+LL | fn f<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
+   |                                    ^^^^^ not a trait
+
+error[E0404]: expected trait, found builtin type `usize`
+  --> $DIR/suggestion-trait-object-issue-139174.rs:9:32
+   |
+LL | fn create_adder<'a>(x: i32) -> usize + 'a {
+   |                                ^^^^^ not a trait
+
+error[E0404]: expected trait, found builtin type `usize`
+  --> $DIR/suggestion-trait-object-issue-139174.rs:16:8
+   |
+LL |     x: usize + 'a,
+   |        ^^^^^ not a trait
+
+error[E0782]: expected a type, found a trait
+  --> $DIR/suggestion-trait-object-issue-139174.rs:3:36
+   |
+LL | fn f<'a>(x: Box<dyn Fn() -> Option<usize + 'a>>) -> usize {
+   |                                    ^^^^^^^^^^
+   |
+help: you can add the `dyn` keyword if you want a trait object
+   |
+LL | fn f<'a>(x: Box<dyn Fn() -> Option<dyn usize + 'a>>) -> usize {
+   |                                    +++
+
+error[E0782]: expected a type, found a trait
+  --> $DIR/suggestion-trait-object-issue-139174.rs:9:32
+   |
+LL | fn create_adder<'a>(x: i32) -> usize + 'a {
+   |                                ^^^^^^^^^^
+   |
+help: `usize + 'a` is dyn-incompatible, use `impl usize + 'a` to return an opaque type, as long as you return a single underlying type
+   |
+LL | fn create_adder<'a>(x: i32) -> impl usize + 'a {
+   |                                ++++
+
+error[E0782]: expected a type, found a trait
+  --> $DIR/suggestion-trait-object-issue-139174.rs:16:8
+   |
+LL |     x: usize + 'a,
+   |        ^^^^^^^^^^
+   |
+help: you can add the `dyn` keyword if you want a trait object
+   |
+LL |     x: dyn usize + 'a,
+   |        +++
+
+error: aborting due to 6 previous errors
+
+Some errors have detailed explanations: E0404, E0782.
+For more information about an error, try `rustc --explain E0404`.

